### ISSUE TITLE
How to check current value of merge_with_ttl_timeout

### ISSUE THREAD

**Issue by ghuname** [ClickHouse/ClickHouse#14737]
Hi,

I have a problem with deleting old records. I know that I can set table's merge_with_ttl_timeout with:
ALTER TABLE [db.]table MODIFY SETTING merge_with_ttl_timeout = 3600

but I would like to check table's current value. I red in documentation that it is set as default to 1 day, but I would like to check it.

1. How to check current value of merge_with_ttl_timeout?

At the moment I am inserting "old" records in a table that has modify setting merge_with_ttl_timeout set to  60 seconds.

2. How to check whether ttl for table has been executed?

3. If something went wrong, how to check it? What events in clickhouse log are relevant for ttl?

Furthermore, it is unclear from which clickhouse version merge_with_ttl_timeout exists.
On https://github.com/ClickHouse/ClickHouse/blob/master/CHANGELOG.md I found mentioning of "merge_with_ttl_timeout" once in ClickHouse release v20.7.2.30-stable, 2020-08-31 as "Corrected merge_with_ttl_timeout logic which did not work well when expiration affected more than one partition over one time interval. (Authored by @excitoon). #12982 (Alexander Kazakov)."

Regards.

**Comment by alesapin** [comment]
Hello!

1) If you change `merge_with_ttl_timeout` value using `ALTER MODIFY SETTING` query than you can check it using `SHOW CREATE TABLE` query in `SETTINGS` section. 

2) When you inserting "fresh" data with "old" values TTL will be executed during the first merge of these parts.

3) It's sad, but in the old versions, we have no information about TTL in logs. In the new versions (starting from 20.9 I think) you can check `merge_type` column in `system.merges` table.

Also, if you are using TTL starting from old clickhouse (< 20.5) versions I recommend running `ALTER MATERIALIZE TTL` query once.

**Comment by ghuname** [comment]
Hi @alesapin, 

1. thanks for your answer. I haven't noticed that show create table shows merge_with_ttl_timeout after altering with modify setting. Thanks.

2. Regarding my test case, I am doing the following:
a) I have created a table something like:

CREATE TABLE database.table
(
    ...
    `ipdrCreationTime` DateTime,
    `insertedOn` DateTime,
    `sign` Int8
)
ENGINE = ReplicatedCollapsingMergeTree('/clickhouse/tables/{layer}-{shard}/table', '{replica}', sign)
PARTITION BY toDate(ipdrCreationTime)
ORDER BY (ipdrCreationTime, ...)
TTL ipdrCreationTime + toIntervalDay(2)
SETTINGS index_granularity = 8192, merge_with_ttl_timeout = 60

b) I am inserting records (one record per day) with datetime from now() to now() - 1 day, now() -2 days... now() - 100 days.

If you are right, during merge, TTL will be executed, so ones per minute old records should be deleted, but this is not happening. In the log I can't find anything useful. I tried with other values for merge_with_ttl_timeout like 600, or 86400, but it is not working on clickhouse server versions:
20.9.1 revision 54439 nor 19.15.4 revision 54426.

If I run OPTIMIZE TABLE database.table FINAL, relevant records are deleted.
It used to work on clickhouse server version  19.16.3 revision 54427.

If it is not working, I don't have any idea how to troubleshoot the problem.

3. I have executed alter table databas.table materialize ttl, but it doesn't help.

At the moment, I have created a bash script that will do optimize table final each night, but I would rather like to use TTL that clickhouse provides if possible.

Best regards.

**Comment by alesapin** [comment]
First of all your schema is quite suboptimal for your data. You have partition key `toDate(ipdrCreationTime)` and if I understand correctly in each insert you have a lot of different days? In this case, each insert will be split into many parts one for each partition and a lot of files will be created on disk.

> If you are right, during merge, TTL will be executed, so ones per minute old records should be deleted

Yes, but only if there is enough space (working threads) in the background pool. If you insert data intensively, then whole background pool can be utilized by normal merges. You can check using selects from `system.merges` table or using `system.metrics`  table (`select * from system.metrics where metric like 'Background%'`).

However, it looks like a bug. Maybe you can provide exact steps to reproduce your case: all queries (create and insert) and ClickHouse version.

**Comment by ghuname** [comment]
> First of all your schema is quite suboptimal for your data. You have partition key `toDate(ipdrCreationTime)` and if I understand correctly in each insert you have a lot of different days? In this case, each insert will be split into many parts one for each partition and a lot of files will be created on disk.

Actually, data are live data, so I will get data for two partitions somewhere around midnight, so it is not a problem.

> > If you are right, during merge, TTL will be executed, so ones per minute old records should be deleted
> 
> Yes, but only if there is enough space (working threads) in the background pool. If you insert data intensively, then whole background pool can be utilized by normal merges. You can check using selects from `system.merges` table or using `system.metrics` table (`select * from system.metrics where metric like 'Background%'`).

As I can see, `select * from system.merges` usually returns no records. I can see records there only if I am doing optimize table... final. Regarding second select, BackgroundSchedulePoolTask = 16, BackgroundPoolTask never goes above 3. So I believe that this is OK.

> However, it looks like a bug. Maybe you can provide exact steps to reproduce your case: all queries (create and insert) and ClickHouse version.

Usecase is very simple. I put part of the table in my reply above as well as versions of clickhouse server. There is nothing special here. It would be useful if clickhouse writes in log what it is trying to do.
For example, if I whould have seen in log something like "Not ehough threads to process TTL (BackgroundPoolTask)" it would be enough to diagnose problem with BackgroundPoolTask threads.

At the moment, I don't have any opportunity to diagnose the issue. 😒

**Comment by tshahmuratov** [comment]
I can confirm it is not working. What I did

1. Created table with data https://clickhouse.tech/docs/ru/sql-reference/statements/alter/ttl/ and filled it same way
2. alter table table_with_ttl modify setting merge_with_ttl_timeout=10;
3. wait for 5 min just in case. Data is still there
4. Ran optimize table table_with_ttl final;
5. Outdated data is dropped

Clickhouse is latest version from docker image without additional settings.
Connected to ClickHouse server version 21.6.4 revision 54448.
